valohai/internals/global_state_loader.py

```diff
@@ -0,0 +1,227 @@
+import argparse
+import json
+import os
+import sys
+from typing import Any, Dict, List, Set, Tuple, Union
+
+from valohai.internals import global_state
+from valohai.internals.input_info import InputInfo
+from valohai.paths import get_inputs_config_path, get_parameters_config_path
+
+
+def load_global_state(
+    default_inputs_from_prepare: Dict = None,
+    default_parameters_from_prepare: Dict = None,
+):
+    """Loads inputs & parameters and stores their value in the global_state
+
+    Inputs & parameters (with values) can be defined in three places:
+    - Call to valohai.prepare()
+    - File-based config in inputs.json and parameters.json
+    - Overridable values from CLI args
+
+    It is possible that something is defined in all the three places simultaneously.
+    Thus, we have a priority order for the final value:
+
+    1. CLI override
+    2. File-based config
+    3. valohai.prepare()
+
+    :param default_inputs_from_prepare: Dict of inputs and their values from valohai.prepare()
+    :param default_parameters_from_prepare: Dict of parameters and their values from valohai.prepare()
+    """
+
+    inputs, parameters = {}, {}
+
+    # Load defaults from call to valohai.prepare()
+    if default_inputs_from_prepare:
+        inputs = sift_defaults(default_inputs_from_prepare)
+    if default_parameters_from_prepare:
+        parameters = sift_defaults(default_parameters_from_prepare)
+
+    # Use inputs.json and parameters.json instead if found
+    file_inputs, file_parameters = load_from_configs()
+    inputs = file_inputs or inputs
+    parameters = file_parameters or parameters
+
+    # Parse and override input & parameter values from CLI
+    cli_inputs, cli_parameters = parse_overrides_from_cli(inputs, parameters)
+    inputs = {key: cli_inputs.get(key, value) for key, value in inputs.items()}
+    final_parameters = {
+        key: cli_parameters.get(key, value) for key, value in parameters.items()
+    }
+
+    # Inputs in various formats are converted into InputInfo(s)
+    final_inputs = {key: convert_to_input_info(value) for key, value in inputs.items()}
+
+    # Save results to global_state
+    global_state.inputs = final_inputs
+    global_state.parameters = final_parameters
+
+    # Store an explicit flag to be used in load_global_state_if_necessary()
+    global_state.loaded = True
+
+
+def flush_global_state():
```

Comment:
Maybe this should be in the `global_state` module? It doesn't seem like a loader.